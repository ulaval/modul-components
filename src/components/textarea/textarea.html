<div class="m-textarea"
     :class="{ 'm--is-disabled' : isDisabled,
               'm--is-waiting' : isWaiting,
               'm--has-error' : hasTextAreaError,
               'm--is-valid' : isTextAreaValid,
               'm--is-focus' : isFocus,
               'm--has-icon' : hasIcon,
               'm--has-label': hasLabel }"
     @click="onClick"
     :style="{ width: inputWidth, maxWidth: inputMaxWidth }">
    <m-input-style :disabled="isDisabled"
                   :waiting="isWaiting"
                   :error="hasTextAreaError"
                   :valid="isTextAreaValid"
                   :label="label"
                   :focus="isFocus"
                   :empty="isEmpty"
                   :icon-name="iconName"
                   :required-marker="requiredMarker"
                   :tag-style="tagStyle">
        <textarea v-m-textarea-auto-height
                  ref="input"
                  class="m-textarea__input"
                  :disabled="!active"
                  :placeholder="placeholder"
                  :readonly="readonly"
                  rows="1"
                  @focus="onFocus"
                  @blur="onBlur"
                  @change="onChange"
                  @keyup="onKeyup"
                  @keydown="onKeydown"
                  @paste="onPaste"
                  v-model="model">
        </textarea>
    </m-input-style>
    <div class="m-textarea__validation">
        <m-validation-message
            class="m-textarea__validation__message"
            :disabled="isDisabled"
            :waiting="isWaiting"
            :error="hasTextAreaError"
            :error-message="errorMessage"
            :valid-message="validMessage"
            :helper-message="helperMessage"></m-validation-message>
        <div class="m-textarea__validation__maxlenght" v-if="hasMaxLength">
            {{ valueLength }}<strong>/{{ maxLength }}</strong>
        </div>
    </div>
</div>
