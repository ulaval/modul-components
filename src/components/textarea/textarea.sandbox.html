<div>
    <h3>
        Test 1 : Compteur de caractères
    </h3>
    <h4>
        Dépassement autorisé
    </h4>
    <ul>
        <li>Nombre de caractères permis : 20</li>
        <li>Affichage du compteur : en tout temps</li>
    </ul>
    <h6>Dépassement autorisé : Oui</h6>
    <br>
    <m-textarea :character-count="true" max-length="20" :length-overflow="true" value="Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat.">
    </m-textarea>

    <h6>Dépassement autorisé : Non</h6>
    <br>
    <m-textarea :character-count="true" max-length="20" :length-overflow="false">
    </m-textarea>

    <h4>
        Seuil d'affichage
    </h4>
    <ul>
        <li>Nombre de caractères permis : 30</li>
        <li>Dépassement autorisé : Oui</li>
    </ul>
    <h6>Affichage du compteur : à partir de 25 caractères</h6>
    <m-textarea :character-count="true" max-length="30" :length-overflow="true" character-count-threshold="25" value="Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat.">
    </m-textarea>

    <h6>Affichage du compteur : à partir de 0 caractère (en tout temps)</h6>
    <br>
    <m-textarea :character-count="true" max-length="30" :length-overflow="true" character-count-threshold="0">
    </m-textarea>

    <h6>Affichage du compteur : à partir de 30 caractères (une fois la limite atteinte)</h6>
    <br>
    <m-textarea :character-count="true" max-length="30" :length-overflow="true" character-count-threshold="30" value="Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat.">
    </m-textarea>

    <h6>Affichage du compteur : à partir de 35 caractères (seuil supérieur à la limite)</h6>
    <br>
    <m-textarea :character-count="true" max-length="30" :length-overflow="true" character-count-threshold="35">
    </m-textarea>

    <h6>Affichage du compteur : à partir de -5 caractères (seuil inférieur à 0)</h6>
    <m-textarea :character-count="true" max-length="30" :length-overflow="true" character-count-threshold="-5">
    </m-textarea>

    <h4>
        Affichage d'un message d'erreur
    </h4>
    <ul>
        <li>Nombre de caractères permis : 20</li>
        <li>Dépassement autorisé : oui</li>
        <li>Affichage du compteur : en tout temps</li>
    </ul>
    <br>
    <m-textarea :character-count="true" max-length="20" :length-overflow="true" value="L'utilisateur a saisi 35 caractères" error-message="Ce champ ne peut comporter plus de 20 caractères.">
    </m-textarea>

    <h4>
        Sous-titre dans un module d'aprentissage (Prisme)
    </h4>
    <ul>
        <li>Nombre de caractères permis : 200</li>
        <li>Dépassement autorisé : oui</li>
        <li>Affichage du compteur : à partir de 180 caractères</li>
    </ul>
    <p class="m-u--typo--precision">176 car. : Lorem ipsum dolor sit amet, consectetur adipiscing elit. Nulla risus ex, dignissim scelerisque vehicula, mempor. Saisir quelques caractères pour faire apparaître le compteur...</p>
    <br>
    <m-textarea placeholder="Je suis un sous-titre" :character-count="true" max-length="200" character-count-threshold="180" max-width="medium"
        :length-overflow="true" v-model="test4Model" :error="test4Model.length > 200 ? true : false">
    </m-textarea>

    <h4>Ne pas afficher un compteur sur un champ de saisie sans limite de caractères</h4>
    <p class="m-u--margin-bottom">
        Lorsque qu'on ajoute un compteur sur un champ sans limite de caractère, le compteur doit être masqué. Nous n'avons pas de cas d'utilisation de compteur de mots sans imposer de limite.
    </p>
    <p>
        <m-textarea :character-count="true" character-counter-threshold="20">
        </m-textarea>
    </p>

    <!-- Sandbox template -->
    <h3>Test # : Title</h3>
    <p class="m-u--margin-bottom">
        Test description
        <!-- if relevant, add link to PR
            </br>From <a href="https://github.com/ulaval/modul-components/pull/000" target="blank">PR description</a>
            -->
    </p>
    <p>
        <m-textarea>
            <!-- Test case -->
        </m-textarea>
    </p>

</div>
