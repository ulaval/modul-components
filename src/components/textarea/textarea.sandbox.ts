import Vue, { PluginObject } from 'vue';
import { Component } from 'vue-property-decorator';

import { TEXTAREA_NAME } from '../component-names';
import WithRender from './textarea.sandbox.html';

@WithRender
@Component
export class MTextareaSandbox extends Vue {
}

const TextareaSandboxPlugin: PluginObject<any> = {
    install(v, options): void {
        v.component(`${TEXTAREA_NAME}-sandbox`, MTextareaSandbox);
    }
};

export default TextareaSandboxPlugin;
