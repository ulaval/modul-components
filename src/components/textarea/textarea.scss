@import 'abstracts/variables';

.m-textarea {
    position: relative;
    display: inline-flex;
    flex-direction: column;
    max-width: $m-max-width--text;

    &.m--has-error {
        .m-textarea {
            &__validation__maxlenght {
                color: $m-color--error;
            }
        }
    }

    &.m--is-valid {
        .m-textarea {
            &__validation__maxlenght {
                color: $m-color--success;
            }
        }
    }

    &__validation {
        display: flex;
        justify-content: flex-end;

        &__message {
            flex: 1 1 auto;
        }

        &__maxlenght {
            transition: color $m-transition-duration ease;
            margin-top: $m-margin--xs;
            font-size: $m-font-size--xs;
        }
    }

    &__input {
        transition: height 2s ease;
        overflow: hidden;
        resize: none;
        width: 100%;
    }
}
