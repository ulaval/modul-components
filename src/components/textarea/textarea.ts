import { ModulVue } from '../../utils/vue/vue';
import { PluginObject } from 'vue';
import Component from 'vue-class-component';
import { Prop, Watch } from 'vue-property-decorator';
import WithRender from './textarea.html?style=./textarea.scss';
import { TEXTAREA_NAME } from '../component-names';
import { InputState } from '../../mixins/input-state/input-state';
import { InputManagement } from '../../mixins/input-management/input-management';
import { KeyCode } from '../../utils/keycode/keycode';
import InputStyle from '../input-style/input-style';
import ValidationMesagePlugin from '../validation-message/validation-message';
import { InputWidth, InputMaxWidth } from '../../mixins/input-width/input-width';

@WithRender
@Component({
    mixins: [
        InputState,
        InputManagement,
        InputWidth
    ]
})
export class MTextarea extends ModulVue {
    @Prop()
    public maxlength: number;

    private internalTextareaError: boolean = false;
    private textareaHeight: string;

    protected mounted(): void {
        this.adjustHeight();
    }

    @Watch('value')
    private valueChanged() {
        this.adjustHeight();
    }

    private get valueLenght(): number {
        let lenght: number = this.as<InputManagement>().internalValue.length;
        if (lenght > this.maxlength) {
            this.internalTextareaError = true;
        } else {
            this.internalTextareaError = false;
        }
        return lenght;
    }

    private get hasMaxlenght(): boolean {
        return this.maxlength > 0;
    }

    private get textareaError(): boolean {
        return this.internalTextareaError || this.as<InputState>().hasError;
    }

    private get textareaValid(): boolean {
        return !this.textareaError && this.as<InputState>().isValid;
    }

    private adjustHeight(): void {
        let el: HTMLElement = (this.$refs.input as HTMLElement);
        el.style.height = 'auto';
        el.style.height = el.scrollHeight + 'px';
    }
}

const TextareaPlugin: PluginObject<any> = {
    install(v, options) {
        console.warn(TEXTAREA_NAME + ' is not ready for production');
        v.use(InputStyle);
        v.use(ValidationMesagePlugin);
        v.component(TEXTAREA_NAME, MTextarea);
    }
};

export default TextareaPlugin;
