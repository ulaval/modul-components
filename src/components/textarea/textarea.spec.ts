import { mount, Wrapper } from '@vue/test-utils';
import Vue from 'vue';

import { renderComponent } from '../../../tests/helpers/render';
import TextareaPlugin from '../../components/textarea/textarea';
import { MTextarea } from './textarea';
import uuid from '../../utils/uuid/uuid';

jest.mock('../../utils/uuid/uuid');
(uuid.generate as jest.Mock).mockReturnValue('uuid');

describe('MTextArea', () => {
    beforeEach(() => {
        Vue.use(TextareaPlugin);
    });

    it('should render correctly', () => {
        const txtarea: Wrapper<MTextarea> = mount(MTextarea);

        return expect(renderComponent(txtarea.vm)).resolves.toMatchSnapshot();
    });

    describe('max-length', () => {
        it('should render correctly state when text length is lesser than max length', () => {
            const txtarea: Wrapper<MTextarea> = mount(MTextarea, {
                propsData: {
                    maxLength: 8,
                    value: '1'
                }
            });

            return expect(
                renderComponent(txtarea.vm)
            ).resolves.toMatchSnapshot();
        });

        it('should render invalid state when text length is greater than max length', () => {
            const txtarea: Wrapper<MTextarea> = mount(MTextarea, {
                propsData: {
                    maxLength: 8,
                    value: '123456789'
                }
            });

            return expect(
                renderComponent(txtarea.vm)
            ).resolves.toMatchSnapshot();
        });
    });
});
