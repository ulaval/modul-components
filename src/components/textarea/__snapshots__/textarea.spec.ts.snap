// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MTextArea max-length should render correctly state when text length is lesser than max length 1`] = `
<div class="m-textarea" style="width:100%;max-width:288px;">
    <div class="m-input-style m--has-value m--is-tag-default" style="width:undefined;">
        <div class="m-input-style__body">
            <div class="m-input-style__body__left">
                <textarea rows="1" class="m-textarea__input">1</textarea>
            </div>
            <div class="m-input-style__body__right"></div>
        </div>
    </div>
    <div class="m-textarea__validation">
        <div class="m-textarea__validation__maxlenght">
            1<strong>/8</strong></div>
    </div>
</div>
`;

exports[`MTextArea max-length should render invalid state when text length is greater than max length 1`] = `
<div class="m-textarea m--has-error" style="width:100%;max-width:288px;">
    <div class="m-input-style m--has-error m--has-value m--is-tag-default" style="width:undefined;">
        <div class="m-input-style__body">
            <div class="m-input-style__body__left">
                <textarea rows="1" class="m-textarea__input">123456789</textarea>
            </div>
            <div class="m-input-style__body__right"></div>
        </div>
    </div>
    <div class="m-textarea__validation">
        <div class="m-textarea__validation__maxlenght">
            9<strong>/8</strong></div>
    </div>
</div>
`;

exports[`MTextArea should render correctly 1`] = `
<div class="m-textarea" style="width:100%;max-width:288px;">
    <div class="m-input-style m--is-tag-default" style="width:undefined;">
        <div class="m-input-style__body">
            <div class="m-input-style__body__left">
                <textarea rows="1" class="m-textarea__input"></textarea>
            </div>
            <div class="m-input-style__body__right"></div>
        </div>
    </div>
    <div class="m-textarea__validation"> </div>
</div>
`;
